GSoC 2018: Changes to MSCX fileformat for better comparability

dmitrio95 opened the pull request:
This pull request contains changes to MSCX fileformat made for better comparability purposes. Some adjustments will probably be needed so this pull request should probably be considered as work-in-progress but I think it can already be presented for reviewing.

## Notes on code changes
### Fileformat changes
Changes directly related to MSCX fileformat are listed [here](https://musescore.org/en/node/271899#comment-841384) so I won't duplicate their description here. I would note though some changes made to the format written in clipboard mode and its differences from MSCX fileformat itself.
1. **Changes in voices storage**
Point 5 of the list posted to the forum says about changes in voices storage introducing the new `<voice>` tag. However this `<voice>` tag is not present in clipboard mode: all job on changing voices is done with `<move>` tag that was also described on the forum.
2. **Changes in making gaps in voices**
Prior to pasting elements to the score a gap should be made by removing/shortening elements that hold the destination position before paste operation. Previously this was [connected](https://github.com/musescore/MuseScore/blob/f77b287a89e4fe9c5e1e6aa0e2869f3e38b06a9f/libmscore/paste.cpp#L151) to reading `<move>` tag which was forced to be written after `<voice>` tags denoting voice offsets. I reworked it so that voice offsets are now defined within `<voiceOffset>` tag and gaps are [made](https://github.com/dmitrio95/MuseScore/blob/scorecmp2/libmscore/paste.cpp#L151) just after all voice offsets are read. This made it also possible to remove a redundant flag passed to `writeSegments` as nothing more depended on it (8fec50175b8a6f3066dad75d378bcaffad271dcf). That way, now a clipboard content looks like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<StaffList version="2.00" tick="0" len="1920" staff="0" staves="1">
  <Staff id="0">
    <voiceOffset>
      <voice id="0">0</voice>
      <voice id="1">0</voice>
      </voiceOffset>
    <Chord>
      <!-- Chord properties -->
      </Chord>
    <!-- other selected elements -->
```
### Code changes
1. **Reading old MSCX files**
As my changes to MSCX format were relatively large I have backed up part of read functions for old fileformat flavor — they are called like `read300()` or `readProperties300()` in the code. Not all such functions were backed up though: `Element::readProperties()` was made to be able to read both old and new fileformat flavor, and `Element::read300()` was defined so that some `Element` subclasses do not require a separate `read300()` function to be defined unless they have a reason (large code change or usage by old fileformat reading code in `read114.cpp` or `read206.cpp`). Such subclasses that do *not* require backing up read functions are:
- All `ScoreElement` subclasses that are not `Element` subclasses;
- All **final** direct subclasses of `Element` class;
- Most of subclasses of `SpannerSegment`.
This change also means that some old reading code can be removed from `read()` functions for most classes but this was not included in this pull request unless was needed for other reasons.
2. **Miscellaneous code changes**
Various code changes were made for solving some minor issues that still were important for my changes to work. Those include changes for more accurate tracking of the current score position by `XmlReader` and `XmlWriter` but there are also other minor changes too.

## Tests and issues with scores reading
Most scores from `mtest` were converted to the new fileformat flavor. Most tests worked after that conversion, others worked after some adjustments to the converted score that were needed since some score files change a bit after load/save cycle. Some examples of such changes may be filling the gaps in the score by rests or adding a number to the tuplet to mark it in the score. However there were some other problems most of which do not affect tests passage but appear not to be a correct behaviour. Below I'll list those problems though I could eventually include in that list some issues that are made by design.

In case someone wants to see real differences to the test scores introduced by a conversion, I would recommend to do it with `git diff -w` command since introduction of `<voice>` tag made most of elements be indented one level more than before.

List of possible problems that were revealed during tests conversion:
1. Final barlines become single on score save if were edited by user (see [this issue](https://musescore.org/en/node/273327)).
2. StaffText elements can move on score save (see [this line of code](https://github.com/musescore/MuseScore/blob/01b96ac8ed924db5452e859aa6a0df535dbb8b85/libmscore/element.cpp#L472)).
3. `<Text>` property of Notes is not handled (example: try to open [this score](https://github.com/musescore/MuseScore/blob/2.3/mtest/guitarpro/artificial-harmonic.gpx-ref.mscx) in MuseScore 2.2 and 3.0).
4. `stepOffset` property of `StaffType` is [written](https://github.com/musescore/MuseScore/blob/4b53934b96849417e6a28f2212aff7a2a858a6ca/libmscore/stafftype.cpp#L210) but is never read.
5. Zero-length spanners. When converting test scores I found lots of zero-length spanners (probably all of them were Voltas). Not sure how they appeared in those scores, and that won't be a problem but it is not even necessary that they are handled and rendered correctly by the existing code. For example, see [this score](https://raw.githubusercontent.com/musescore/MuseScore/fe8a3e5e9c6430efa65d9e6338e4d1dd326ff9cd/mtest/libmscore/layout/goldberg.mscx), Volta with `id="396"`. This lucky volta appears to be just at the boundary between two systems (page 31, end of the first system should be the place). As a result, it is not handled during layout (as does not pass [this check](https://github.com/musescore/MuseScore/blob/dff4fcc65fa3a2b5abcc67accc0d879d12c91450/libmscore/layout.cpp#L3302)), is not displayed and has no start and end elements. A similar check can be found in the same file above ([here](https://github.com/musescore/MuseScore/blob/dff4fcc65fa3a2b5abcc67accc0d879d12c91450/libmscore/layout.cpp#L3210)), and correcting the previously mentioned check to be the same as this one could fix the problem, but are zero-length voltas something that was added by design?
6. Various tags not handled (though these may be not handled by design, not sure): `<xoffset>`, `<yoffset>` within `<Tempo>`; `<pedalY>` (handled for old scores only), `<trillY>` (not handled at all).
7. Not sure it is a problem but still: beams properties (`<l1>` , `<l2>`) in [mtest/libmscore/tools/undoSlashRhythm01-ref.mscx](https://github.com/musescore/MuseScore/blob/5e116fe595e37ef6ca496965dbf64047e21951d9/mtest/libmscore/tools/undoSlashRhythm01-ref.mscx) change after loading and saving a score in test mode.

shoogle commented:
@dmitrio95, great job! Thanks for submitting this. It's going to take a while to review, especially with 2.3 going out in the coming week, so relax this afternoon and read up on XML schema in case we decide to go down that route, and then make a new branch tomorrow to start working on the textual diff viewer. The viewer should be a translatable bullet-point list of differences between two files, but there should also be an option to view the raw MSCX diff. Discussion should take place on the MuseScore forums.

Regarding the XML schema, do the W3Schools tutorials on [XSD Schema](https://www.w3schools.com/xml/schema_intro.asp), [XSD Complex](https://www.w3schools.com/xml/schema_complex.asp) and [XSD Data](https://www.w3schools.com/xml/schema_dtypes_string.asp). Don't try to learn it off by heart; just give it a quick read to try to get a feeling for what sort of things are possible. The main purpose is obviously to validate files, but one cool thing a schema allows you to do is create a [visualisation of the format](https://stackoverflow.com/questions/2486758/how-to-visualize-an-xml-schema), document the format, or even parse the format and [convert between formats](https://www.musicxml.com/for-developers/musicxml-xslt/) without having a write a parser in each new language you want to port it to (you write the parser once in XLS and import that using an XLST library for the chosen language). None of these are essential for MSCX, but what would be cool is to see how the format evolves over time by comparing different versions of the schema. It might also encourage better practise when it comes to adding more features to MSCX.

Jojo-Schmitz commented:
There's a rebase needed

shoogle commented on libmscore/point.h:
```diff
@@ -0,0 +1,76 @@
+class XmlReader;
+class XmlWriter;
+
+class PointInfo {
```
@dmitrio95, maybe this class should be called `Move` or `UniveralMove` or something to make it clear that it corresponds to `<move>` in the MSCX?

dmitrio95 replied:
I support the idea on renaming both this class and the corresponding tag to something more consistent. For cross-staff chords I would rather return the old syntax (`<move>1</move>`) as it would be more consistent with other elements' properties which are normally written in separate tags which are not mixed with anything else. But this doesn't matter so much.

Concerning the name of the renamed tag, a `<pos>` tag denotes now an element offset in a score. Although I don't know reasons for that (there is an `<offset>` tag that denotes the same property) it can potentially be not too convenient to have both `<pos>` and `<position>` tags. So I think we either have to always use `<offset>` instead of `<pos>` or prefer `<location>` name for the renamed `<move>` tag.

dmitrio95 commented on libmscore/segment.h:
```diff
@@ -187,7 +187,8 @@ class Segment final : public Element {
       virtual int rtick() const override         { return _tick;  } // tickposition relative to measure start
-      Fraction fpos() const;
+      Fraction rfrac() const;
+      Fraction afrac() const;
```
I would also note that a similar `Element::ftick` function appeared in MuseScore code during the recent time:
https://github.com/musescore/MuseScore/blob/10ebbb9080d69bb3629ca48dfb76eb2a5aadef9e/libmscore/element.h#L366
I think I will merge it with those `afrac` functions introduced in my branch after performing a rebase.

dmitrio95 commented on libmscore/point.cpp:
```diff
@@ -0,0 +1,252 @@
+            xml.tag("notes", _note, relDefaults._note);
+            xml.etag();
+            }
+      else {
```
No, in fact, it isn't used anywhere. Perhaps it is really better to remove it but I am not sure about that.

shoogle replied:
I am happy for you to keep the ability for a PointInfo to be absolute *within the code* as that might come in handy, but I don't want to see `<move_abs>` in any MSCX files as it will spoil the diff. You could do something like `Q_ASSERT(isRelative())` in one of the save functions.

dmitrio95 replied:
OK, I'll remove that `<move_abs>`-related code then.

dmitrio95 commented on libmscore/connector.cpp:
```diff
@@ -0,0 +1,447 @@
+            return false;
+      if (_type != other->_type)
+            return false;
+      if (!_prev && hasPrevious() && !other->_next && other->hasNext()) {
```
Well, this looks better, thanks. I'll change this and similar checks then.

dmitrio95 commented on libmscore/connector.cpp:
```diff
@@ -0,0 +1,447 @@
+
+void ConnectorInfo::forceConnect(ConnectorInfo* other)
+      {
+      if (!other || (this == other))
```
But those `Q_ASSERT`s are stripped from release builds, I guess?
That says also the [developers handbook](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules#Qt-Debugging-constructs):
`Q_ASSERT(bool test) should be used for errors for which musescore can still continue on. Q_ASSERT does nothing when built in release, but will halt execution when built in debug.`

This check covers cases which are hardly acceptable for MuseScore runtime: a null pointer would lead to a crash and `this == other` would cause an eternal loop when something in the code would try to find an endpoint of such a connector. Perhaps there could be a warning added but I would not replace this check with `Q_ASSERT`.

shoogle replied:
Ok cool. I just wondered whether you might be able to catch this before the function is called, but if you can't be sure of doing that then this is fine.

shoogle commented on libmscore/connector.h:
```diff
@@ -0,0 +1,146 @@
+      bool hasPrevious() const      { return (_prevInfo.measure() != INT_MIN); }
+      bool hasNext() const          { return (_nextInfo.measure() != INT_MIN); }
+      bool isStart() const          { return (!hasPrevious() && hasNext()); }
+      bool isMiddle() const         { return (hasPrevious() && hasNext());  }
```
OK, cool. Well it can stay as far as I am concerned, though you can remove it or `Q_ASSERT(!isMiddle())` if you think it will cause problems.

dmitrio95 commented on libmscore/connector.h:
```diff
@@ -0,0 +1,146 @@
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2018 Werner Schweer
```
As far as I remember, that connectors-related classes code is what I wrote entirely myself. Anyway, I am happy with this code being licensed under GPL-2. Concerning the copyright notice itself, it doesn't really matter for me, I just made it consistent with other parts of MuseScore code. Perhaps something like "MuseScore contributors" would be more precise but as no other files uses any similar statements then I would let it be as is.

dmitrio95 commented:
I have just updated this pull request and fixed those issues that were discussed above. I have also renamed `<move>` tag to `<location>` in most contexts and returned an old syntax for writing `staff move` property which now uses `<staffMove>` tag. Those changes are also briefly described [on the forum](https://musescore.org/en/node/271899#comment-848028).

Jojo-Schmitz commented:
One seemingly legit mtest fail on Travis CI: https://travis-ci.org/musescore/MuseScore/jobs/414767245#L5568

shoogle commented:
@lasconic, @anatoly-os, I consider this ready to merge. It would be great if this could happen ASAP so that @dmitrio95 can submit a second PR for the UI without including these commits again in that PR.

(@dmitrio95, feel free to submit a second PR anyway and rebase it one this one is merged.)

anatoly-os commented:
I would squash most of the commits and keep the most important and general ones in the history. Also, I want @wschweer to have a look at this.

anatoly-os commented:
@dmitrio95 this PR looks good to merge. Once you squash the commits, we will merge it. After that, you will need to create the PR removing read300 functionality proposed by @wschweer.

shoogle commented:
__Question:__ is backward compatibility for 2.x separate to the 3.x functions?

__Comment:__ there should probably be at least one nightly build that can read and write both formats (old and new 3.x), so the method proposed by @anatoly-os (i.e. a separate PR to remove the old format) is sensible.

anatoly-os commented:
@dmitrio95 are you around? Because of the scale of the changes in this PR, we need to merge it asap to prevent further rebasing/merging issues in other PRs and commits from core developers.

dmitrio95 commented:
@anatoly-os yes, I have reduced this branch to about 30 commits for now, and I guess after reducing them a bit more and checking that all works I will be able to push them. This will be no further than tomorrow.

Still I would like to ask whether it is OK to push the squashed commits right here or I am better to create a separate "squashed" PR. This pull request is a part of my GSoC final report, and it will be not too well if this PR will not be counted just because of the rewritten history missing the GSoC's deadline.

dmitrio95 commented:
Concerning @shoogle's question on compatibility functions, I used `read300` functions in some parts of 1.x and 2.x formats reading code so it would probably be not possible to remove all of `read300` functions. Still after dropping old 3.0 format there should be less of them and they could be moved away from element classes. I'll investigate this once I am finished with this PR.

anatoly-os commented:
@dmitrio95 you can create separate PR and keep this one opened for GSoC evaluation.

dmitrio95 commented:
As #3870 has been merged there seems to be no reason for having this pull request open anymore. I guess it can be closed.
